### PR TITLE
[RFC][UN-14114] Add groups to uni select

### PR DESCRIPTION
**Pull request by carantunes**
https://uniplaces.atlassian.net/browse/UN-14114

screen shot 2018-08-30 at 09 52 21

### CODE REVIEW

File: addon/components/uni-select.js [hunk 1]
```diff
@@ -10,9 +11,15 @@ export default Component.extend({
 
   options: [],
   selected: null,
+  selectedGroup: null,
   placeholder: null,
   useAlias: false,
   aliasValue: null,
+  groups: [],
+
+  hasGroups: computed('groups', function() {
```
Comment (AndreJoaquim) [inline addon/components/uni-select.js]: You can use `computed.gt('groups.length', 0)`

**Review by AndreJoaquim — COMMENTED** [review]
After already talking with @carantunes, I don't think that this is not the best solution for this problem. I think that it's not good to be able to have two separate APIs for the same component -- `groups` vs. `options`.

One of the problems is the following: what will happen when someone invokes `{{uni-select groups=groups options=options}}`?

Also, this solution doesn't allow to create a `uni-select` with groups and regular options under the same `<select>`.

My suggestion, and the one I believe is the most correct solution, is having an optional field inside the option object passed to the uni-select component:

```js
const selectOptions: [{
    label: ‘this is a group’,
    options: [] // If it has options, it'll behave as a group
  },
  {
    label: 'this is a regular option'
  }
]
```
and invoking the component as we already do:
```
{{uni-select options=selectOptions}}
```

This will lead to:
- Not having possible conflict situations: groups and options parameters both being passed
- Keeping the same API: when we want to add groups to already existent uni-select, it's adding the `options` array
- Allowing `optgroup`s and `option`s to co-exist

**Comment by carantunes** [comment]
I don't think its a good behaviour to have both options alone and groups in the same select. Its confusing to select an option without group, and not being able to select a group. 

Having that said and as this is an improvement to the code and as there's no other select with groups in the works, i suggest creating a ticket for the frontend guild and i can tackle it in the next sprint.

File: addon/components/uni-select.js [hunk 2]
```diff
@@ -24,36 +31,50 @@ export default Component.extend({
...
+      let group = this.get('hasGroups') ? this.get('groups')[0].key : null;
+      this._changeAliasValue(this._getFirstAvailableValue(), group);
     }
   },
 
   actions: {
     changeSelected({ target }) {
       if (this.get('useAlias')) {
-        this._changeAliasValue(target.value);
+        let group = this.get('hasGroups') ? target.options[target.selectedIndex].parentNode.getAttribute('key') : null;
```
Comment (henrmota) [inline addon/components/uni-select.js:46]: this has more than 120 chars?

File: addon/components/uni-select.js [hunk 3]
```diff
@@ -24,36 +31,50 @@ export default Component.extend({
...
       }
 
       this.get('onChange')(target.value);
     }
   },
 
-  _changeAliasValue(key) {
-    let option = A(this.get('options')).findBy('key', key);
+  _changeAliasValue(key, group = null) {
+    let options = this.get('options');
```
Comment (henrmota) [inline addon/components/uni-select.js]:
```JS
let options = group ? A(this.get('groups')).findBy('key', group).options : this.get('options');
```

**Review by henrmota — CHANGES_REQUESTED** [review]
🔝

**Review by henrmota — APPROVED** [review]

**Review by migueldemoura — APPROVED** [review]

File: addon/components/uni-select.js (hunk 2 above)
Comment (carantunes) [reply addon/components/uni-select.js:46]: exactly 120 😆

**Review by mmelo — APPROVED** [review]

**Review by mmelo — APPROVED** [review]

**Pull request merged**